### PR TITLE
add support for master/worker combined nodes

### PR DESCRIPTION
**Pull request by jensfr**
In some compact clusters variants nodes have both the master and worker
role.  However, the machine config pools that are used in
these configurations have all nodes in the master pool and zero nodes in
the worker pool.

Zero nodes in the worker pool breaks our current practice of installing
kata on all nodes in the worker pool. The solution is to install on all
nodes in the master pool instead. For that we fetch the worker pool
information and check if it has nodes in it. If it has a machine count
of zero we use the master pool instead.

This has an important implication:
in this master-only configuration we can't support installing on a
subset of nodes by setting a label on the selected nodes the user wants
to install kata on. Why? This feature depends on creating a custom pool
that inherits from the worker pool. MCP only supports inheriting from
the worker pool but not from the master pool. See
https://github.com/openshift/machine-config-operator/blob/master/docs/custom-pools.md

This is a trade-off for being able to support these compact cluster
types.

Signed-off-by: Jens Freimann <jfreimann@redhat.com>

### CODE REVIEW

**Comment by jensfr** [comment]
somethiing wrong with this, checking...

**Comment by harche** [comment]
@jensfr have you tested this on a regular 6 nodes OC cluster?

**Comment by jensfr** [comment]
@harche Yes, I've tested it also on regular 6 node clusters in the cloud and on 5-node  configuration on my bare-metal cluster.

File: pkg/controller/kataconfig/openshift_reconciler.go [hunk 1]
```diff
@@ -405,22 +412,61 @@ func (r *ReconcileKataConfigOpenShift) listKataPods() error {
 	return nil
 }
 
+func kataOcExists(r *ReconcileKataConfigOpenShift) (bool, error) {
```
Comment (harche) [inline pkg/controller/kataconfig/openshift_reconciler.go:415]: if you change the method signature to `func ((r *ReconcileKataConfigOpenShift) kataOcExists()` you don't have to pass a pointer to the object as an argument.

File: pkg/controller/kataconfig/openshift_reconciler.go [hunk 2]
```diff
@@ -405,22 +412,61 @@ func (r *ReconcileKataConfigOpenShift) listKataPods() error {
...
+	kataOcMcp := &mcfgv1.MachineConfigPool{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: "kata-oc"}, kataOcMcp)
+	if err != nil && errors.IsNotFound(err) {
+		r.kataLogger.Info("No kata-oc machine config pool found!")
```
Comment (harche) [inline pkg/controller/kataconfig/openshift_reconciler.go:419]: I haven't checked, but I was wondering does this logger not have `Error` or do we must use `Info` to convey the error message?

File: pkg/controller/kataconfig/openshift_reconciler.go [hunk 3]
```diff
@@ -405,22 +412,61 @@ func (r *ReconcileKataConfigOpenShift) listKataPods() error {
...
+		return false, nil
+	} else if err != nil {
+		r.kataLogger.Info("Something went wrong trying to get the kata-oc machine config pool!")
+		return false, err
+	}
+
+	return true, nil
+}
+
+func workerOrMaster(r *ReconcileKataConfigOpenShift) (string, error) {
```
Comment (harche) [inline pkg/controller/kataconfig/openshift_reconciler.go:429]: same as, https://github.com/openshift/kata-operator/pull/17/files#r509025406

**Review by harche — COMMENTED** [review]
minor comments, otherwise LGTM

File: pkg/controller/kataconfig/openshift_reconciler.go (hunk 1 above)
Comment (jensfr) [reply pkg/controller/kataconfig/openshift_reconciler.go:415]: good point :-) I'll do that, thx

File: pkg/controller/kataconfig/openshift_reconciler.go (hunk 2 above)
Comment (jensfr) [reply pkg/controller/kataconfig/openshift_reconciler.go]: It does have Error, I changed it here and in a few other places added in this PR. Thanks!

File: pkg/controller/kataconfig/openshift_reconciler.go (hunk 3 above)
Comment (jensfr) [reply pkg/controller/kataconfig/openshift_reconciler.go]: fixed, thanks!

**Pull request merged**